### PR TITLE
Train and predict

### PR DESCRIPTION
**Pull request by finn-freiheit**

### CODE REVIEW

File: predict.ipynb [hunk 1]
```diff
@@ -0,0 +1,220 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "predictions = model.predict(\n",
+    "    ds.map(util.parse_image).batch(32),\n",
+    "    # docu says that \"do not specify the batch_size is your data is in the form of .., dataset, ...\"\n",
```
Comment (vladhc) [inline predict.ipynb]: typo? "... *if* you data is ..."

File: predict.ipynb [hunk 2]
```diff
@@ -0,0 +1,220 @@
...
+    "    # but it must be explizitly set to \"None\" even if the docu says the default values is \"None\". Otherwise there will be an error.\n",
```
Comment (vladhc) [inline predict.ipynb]: typo? *explicitly*
I haven't got the meaning two-line comment.

File: predict.ipynb [hunk 3]
```diff
@@ -0,0 +1,220 @@
...
+    "df_predictions.head()"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## plot predicted images with classification\n",
+    "\n",
+    "We plot a few images for a brief sanity check by the scientists. The prediction is shown in the external as well as in the internal nomenclature. Also, the confidence of the network for the classification is shown."
```
Comment (vladhc) [inline predict.ipynb]: Why "by the scientists"!? :) It's for everybody!

File: train.ipynb [hunk 4]
```diff
@@ -0,0 +1,499 @@
+{
+ "cells": [
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "# Training\n",
+    "\n",
+    "In this notebook, we will load the training images and training classification and export them to `tfrecords` files. Then we will use the new Keras high-level API to train a model and save the model to disk. We will use this model in the `predict.ipynb` notebook to predict the class of unclassified images.\n",
```
Comment (vladhc) [inline train.ipynb]: Please, shorter. Some words don't bring any profit, only make it harder to grasp. Can be removed:
* "In this notebook"
* "use the new Keras high-level API to",
* "and save the model to disk"
* "to predict the class of unclassified images"

File: train.ipynb [hunk 5]
```diff
@@ -0,0 +1,499 @@
...
+    "use_class_weight = True"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## load classification\n",
+    "\n",
+    "In this section, the notebook loads the classification of the training images into a training and an evaluating data frame. The notebook expects that there is one classification file per folder with training images. The folder must have the same name as the classification file without the file extension. The images in the folder must be named exactly like in the classification file. There can be multiple classification files and folders."
```
Comment (vladhc) [inline train.ipynb]: Shorter:
* "In this section, the notebook loads " -> "Load"
* "The notebook expects that there is " -> "There should be"

File: train.ipynb [hunk 6]
```diff
@@ -0,0 +1,499 @@
...
+   "source": [
+    "df_show_balance = df_train.groupby(\"int_label\").count()\n",
+    "print(df_show_balance)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "The training data set is heavily unbalanced. The make the prediction also accurate for the minority classes, one can use a weighted loss function. If `use_class_weight`, we will calculate a weight for each class.\n",
```
Comment (vladhc) [inline train.ipynb]: Short and clean! :+1: :)
Is there a typo? "The make the ..." -> "To make the ..."; "prediction" -> "predictions"

File: train.ipynb [hunk 7]
```diff
@@ -0,0 +1,499 @@
...
+    "class_weight"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## exporting the data to tfrecods files\n",
+    "\n",
+    "Below, we will export the training and evaluating data set to tfrecods files. We will use them to train the network and get the best weights for our model."
```
Comment (vladhc) [inline train.ipynb]: "Below", "and get the best weights for our model" can be removed. Nobody trains to get the worst weights :)

File: train.ipynb [hunk 8]
```diff
@@ -0,0 +1,499 @@
...
+    "util.export_dataframe(df_train, cfg.path_to_train_data_set)"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## training of the model\n",
+    "\n",
+    "Here we are training the model. The training can be observed by using a TensorBoard. The model will check the evaluation loss and accuracy after each epoch. It will save the model weights if the evaluation loss is smaller than the previous one. After the training is done, the best weights are loaded and the model is saved to disk.\n",
```
Comment (vladhc) [inline train.ipynb]: should we notice how to use TensorBoard in order to observe the training?
"The model will check the evaluation loss " -> "Keras will check the evaluation loss"

File: train.ipynb [hunk 9]
```diff
@@ -0,0 +1,499 @@
...
+    "\n",
+    "**Important:** We will use this file in the `predict.ipynb`. If the training is done in the cloud and the prediction should be done on a local machine, the `model.h5` must be transferred."
```
Comment (vladhc) [inline train.ipynb]: transferred where to?

File: train.ipynb [hunk 10]
```diff
@@ -0,0 +1,499 @@
...
+    "    tf.keras.layers.Dropout(rate=0.2),\n",
+    "    tf.keras.layers.Dense(64, activation='relu'),\n",
+    "    tf.keras.layers.Dropout(rate=0.2),\n",
+    "    tf.keras.layers.Dense(6, activation='softmax', name='predictions')\n",
+    "])\n",
+    "\n",
+    "# default values for adam optimizer\n",
+    "# keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False)\n",
+    "# see: https://keras.io/optimizers/\n",
+    "adam = tf.keras.optimizers.Adam(lr=0.001, beta_1=0.9, beta_2=0.999, epsilon=None, decay=0.0, amsgrad=False)\n",
```
Comment (vladhc) [inline train.ipynb]: why to specify default values and also give them here?

File: train.ipynb [hunk 11]
```diff
@@ -0,0 +1,499 @@
...
+  },
+  {
+   "cell_type": "code",
+   "execution_count": null,
+   "metadata": {},
+   "outputs": [],
+   "source": [
+    "predictions = model.predict(\n",
+    "    tf.data.TFRecordDataset(cfg.path_to_eval_data).map(util.parse_tfrecord).batch(32),\n",
+    "    # docu says that \"do not specify the batch_size is your data is in the form of .., dataset, ...\"\n",
```
Comment (vladhc) [inline train.ipynb]: looks like copy-paste, see comment above for the same code

File: train.ipynb [hunk 12]
```diff
@@ -0,0 +1,499 @@
...
+    "    print(\"accuracy for class {}: {}\".format(i, hits/all_of_class))"
+   ]
+  },
+  {
+   "cell_type": "markdown",
+   "metadata": {},
+   "source": [
+    "## plot the certainty per class\n",
+    "\n",
+    "We have to check the certainty overall distribution and the certainty distribution of the individual classes. If there is a correlation between wrongly classified images and a low certainty of the network, we could use this as a selector for candidates for a double check by humans."
```
Comment (vladhc) [inline train.ipynb]: "overall" -> "over all" ?

File: bodestm/util.py [hunk 13]
```diff
@@ -0,0 +1,167 @@
...
+from collections import defaultdict
+import tensorflow as tf
+import glob
+import numpy as np
+import gitignored.remap as remap
+
+
+def bytes_feature(value):
+    """Returns a bytes_list from a string / byte."""
+    if isinstance(value, type(tf.constant(0))):
```
[the remark this answers is not in the document]
Comment (finn-freiheit) [reply bodestm/util.py]: removed lines

File: bodestm/util.py [hunk 14]
```diff
@@ -0,0 +1,167 @@
...
+    image = tf.image.decode_png(parsed['image_raw'], channels=1)  # output as grayscale
+    image = tf.image.convert_image_dtype(image, tf.float32)
+    return image, parsed["int_label"]
+
+
+def load_classification(path_to_classification):
+    df = pd.read_csv(path_to_classification, names=["file_name", "label"], delimiter="\t", header=None, dtype=str)
+    df["label"] = df["label"].astype(int)
+
+    df.insert(len(df.columns), "generated", False)
```
[the remark this answers is not in the document]
Comment (finn-freiheit) [reply bodestm/util.py]: yes -> done

File: bodestm/util.py [hunk 15]
```diff
@@ -0,0 +1,167 @@
...
+
+    measurement_series = []
+    for file_name in df["file_name"]:
+        measurement_series.append(
+            "{}_{}_{}".format(file_name.split("_")[0], file_name.split("_")[1], file_name.split("_")[2]))
```
[the remark this answers is not in the document]
Comment (finn-freiheit) [reply bodestm/util.py]: this function is deprecated / unused -> I removed it

File: bodestm/util.py [hunk 16]
```diff
@@ -0,0 +1,167 @@
...
+
+    measurement_index = []
+    for file_name in df["file_name"]:
+        measurement_index.append(file_name.split("_")[-1].split(".")[0])
```
[the remark this answers is not in the document]
Comment (finn-freiheit) [reply bodestm/util.py]: this function is deprecated / unused -> I removed it

File: bodestm/util.py [hunk 17]
```diff
@@ -0,0 +1,167 @@
...
+
+    df["measurement_series"] = measurement_series
```
[the remark this answers is not in the document]
Comment (finn-freiheit) [reply bodestm/util.py]: this function is deprecated / unused -> I removed it

File: train.ipynb (hunk 4 above)
Comment (finn-freiheit) [reply train.ipynb]: shortened

File: train.ipynb (hunk 5 above)
Comment (finn-freiheit) [reply train.ipynb]: done

File: train.ipynb (hunk 6 above)
Comment (finn-freiheit) [reply train.ipynb]: done :-)

File: train.ipynb (hunk 7 above)
Comment (finn-freiheit) [reply train.ipynb]: removed the whole sentence

File: train.ipynb (hunk 10 above)
Comment (finn-freiheit) [reply train.ipynb]: I am now using the default values. But I want to play around with them to increase the accuracy. Then I want to have the default values as a template.

File: train.ipynb (hunk 11 above)
Comment (finn-freiheit) [reply train.ipynb]: removed comment

File: train.ipynb (hunk 12 above)
Comment (finn-freiheit) [reply train.ipynb]: changed to "overall certainty"

File: train.ipynb (hunk 9 above)
Comment (finn-freiheit) [reply train.ipynb]: shortened:

**Important:** Make sure to use the most recent `model.h5` file.

File: train.ipynb (hunk 8 above)
Comment (finn-freiheit) [reply train.ipynb]: I think explaning how to use TensorBoard would be too verbose here. I changed "The model" to "Keras".

File: bodestm/util.py [hunk 18]
```diff
@@ -0,0 +1,167 @@
...
+    parsed = tf.io.parse_single_example(record, {
+        'image_raw': tf.io.FixedLenFeature((), tf.string),
+        'int_label': tf.io.FixedLenFeature((), tf.int64),
+    })
+    image = tf.image.decode_png(parsed['image_raw'], channels=1)  # output as grayscale
+    image = tf.image.convert_image_dtype(image, tf.float32)
+    return image, parsed["int_label"]
+
+
+def load_classification(path_to_classification):
```
[the remark this answers is not in the document]
Comment (finn-freiheit) [reply bodestm/util.py]: this function is deprecated / unused -> I removed it

File: bodestm/util.py [hunk 19]
```diff
@@ -151,17 +123,54 @@ def get_classification_df(classification_files):
...
-    df["int_label"] = remap.remap_cross(df.ext_label)
+    df["int_label"] = remap_cross(df["ext_label"])
         
     return df
+
+def remap_cross(s):
+    s_remap = s.copy()
+    for i in range(len(s_remap.values)):
+        element = s_remap.values[i]
+        if element == 0:
```
Comment (vladhc) [inline bodestm/util.py]: this multiline `if` can be expressed a bit better via dictionary:
```
buf = {
  0: 0,
  1: 1,
  3: 2,
  4: 3,
  5: 4,
  6: 5,
}[element]
```

File: bodestm/util.py [hunk 20]
```diff
@@ -151,17 +123,54 @@ def get_classification_df(classification_files):
...
+        elif element == 6:
+            buf = 5
+        s_remap[i] = buf
+    
+    return s_remap
+    
+def remap_cross_back(s):
+    s_remap = s.copy()
+    for i in range(len(s_remap)):
+        element = s_remap.values[i]
```
Comment (vladhc) [inline bodestm/util.py]: same as above regarding multiline `if`

File: bodestm/util.py (hunk 19 above)
Comment (finn-freiheit) [reply bodestm/util.py]: done

File: bodestm/util.py (hunk 20 above)
Comment (finn-freiheit) [reply bodestm/util.py]: done

File: bodestm/util.py [hunk 21]
```diff
@@ -151,17 +123,48 @@ def get_classification_df(classification_files):
...
+    
+    s_remap = s.copy()
+    for i in range(len(s_remap.values)):
+        element = s_remap.values[i]
+        s_remap[i] = buf[element]
+    
+    return s_remap
+    
+def remap_cross_back(s):
+    buf = {
```
Comment (vladhc) [inline bodestm/util.py]: Sorry, now I see that this mapping can be done even cleaner:
1. make a constant `LABEL_REMAPPING = { ...fill dictionary values here from line 138.. }`
2. make a constant `LABEL_REMAPPING_REVERSE = {target: src for src, target from LABEL_REMAPPING.items()}`
3. use these constants in `remap_cross_back()` and `remap_cross()`
Profit! :)

This way we avoid possible mistakes in inconsistency (when one side mapping is 3->2, but reverse due to mistake is 2->1).

Comment (finn-freiheit) [reply bodestm/util.py]: cool trick! way better!
